haskell-src/Concordium/Types/Parameters.hs

```diff
@@ -27,42 +397,313 @@ import qualified Concordium.Crypto.SHA256 as Hash
 import Concordium.ID.Parameters
 import Concordium.Types
 import Concordium.Types.HashableTo
-import Concordium.Utils
 
 -- |Chain cryptographic parameters.
 type CryptographicParameters = GlobalContext
 
--- |The mint-per-slot rate is part of the 'MintDistribution' parameters for 'ChainParametersV0',
--- but as of 'ChainParametersV1' is replaced by a mint-per-payday rate in the 'TimeParameters'.
--- 'MintPerSlotForCPV0'
-data MintPerSlotForCPV0 cpv where
-    MintPerSlotForCPV0Some :: {_mpsMintPerSlot :: !MintRate} -> MintPerSlotForCPV0 'ChainParametersV0
-    MintPerSlotForCPV0None :: MintPerSlotForCPV0 'ChainParametersV1
-
-instance IsChainParametersVersion cpv => Serialize (MintPerSlotForCPV0 cpv) where
-    put MintPerSlotForCPV0Some{..} = put _mpsMintPerSlot
-    put MintPerSlotForCPV0None = return ()
-    get = case chainParametersVersion @cpv of
-        SCPV0 -> MintPerSlotForCPV0Some <$> get
-        SCPV1 -> return MintPerSlotForCPV0None
-
--- |Lens for '_mpsMintPerSlot'
-{-# INLINE mpsMintPerSlot #-}
-mpsMintPerSlot :: Lens' (MintPerSlotForCPV0 'ChainParametersV0) MintRate
-mpsMintPerSlot =
-    lens _mpsMintPerSlot (\mps x -> mps{_mpsMintPerSlot = x})
-
-deriving instance Eq (MintPerSlotForCPV0 cpv)
-deriving instance Show (MintPerSlotForCPV0 cpv)
+-- The Template Haskell below generates the quoted definitions, together with the singletons
+-- and liftings to singletons. The module header is used to give the haddock documentation.
```

Comment:
Explain what exactly is being generated, e.g., what is a "quoted definition"?

Reply:
Elaborated.